File: leetcode/implement-trie-prefix-tree/implement-trie-prefix-tree.cs
// LeetCode #208 - Implement Trie (Prefix Tree)
// https://leetcode.com/problems/implement-trie-prefix-tree

public class Trie {
    public void Insert(string word)
    {
        var node = _root;

        foreach (var ch in word) {
            if (node.TryGetValue(ch, out var child)) {
                node = child;
            } else {
                child = new Node();
                node.Add(ch, child);
                node = child;
            }
        }

        node.IsWord = true;
    }

    public bool Search(string word)
        => Traverse(word) is Node node && node.IsWord;

    public bool StartsWith(string prefix)
        => Traverse(prefix) is Node;

    private sealed class Node : Dictionary<char, Node> {
        internal bool IsWord { get; set; } = false;
    }

    private Node Traverse(string prefix)
    {
        var node = _root;

        foreach (var ch in prefix) {
            if (!node.TryGetValue(ch, out node)) return null;
        }

        return node;
    }

    private readonly Node _root = new Node();
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */

File: leetcode/implement-trie-prefix-tree/implement-trie-prefix-tree.java
// LeetCode #208 - Implement Trie (Prefix Tree)
// https://leetcode.com/problems/implement-trie-prefix-tree

class Trie {
    public void insert(String word) {
        var node = _root;

        for (var i = 0; i < word.length(); ++i)
            node = node.computeIfAbsent(word.charAt(i), k -> new Node());

        node.isWord = true;
    }

    public boolean search(String word) {
        var node = traverse(word);
        return node != null && node.isWord;
    }

    public boolean startsWith(String prefix) {
        return traverse(prefix) != null;
    }

    private static class Node extends HashMap<Character, Node> {
        boolean isWord = false;
    }

    private Node traverse(String prefix) {
        var node = _root;

        for (var i = 0; i < prefix.length(); ++i) {
            node = node.get(prefix.charAt(i));
            if (node == null) return null;
        }

        return node;
    }

    private final Node _root = new Node();
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
